### PR TITLE
Basic commands for configuring logging subsystem have been implemented.

### PR DESCRIPTION
**Pull request by istraka**

### CODE REVIEW

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandler.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Batch;
+import org.wildfly.extras.creaper.core.online.operations.OperationException;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.Values;
+
+/**
+ * @author Ivan Straka istraka@redhat.com
+ */
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandler.java]: Please avoid the `@author` tags, this information is available from Git history.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandler.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+                .andOptional("encoding", encoding.displayName())
+                .andOptional("target", target == null ? null : target.value())
+                .andOptional("named-formatter", namedFormatter)
+                .andOptional("formatter", patternFormatter);
+
+
+        Batch batch = new Batch();
+        batch.add(handlerAddress, values);
+
+        ops.batch(batch);
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandler.java]: Why a batch when you're only adding a _single_ operation to the batch?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandler.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+
+    public static final class Builder extends ManipulateConsoleHandler.Builder<Builder> {
+
+        private boolean replaceExisting = false;
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandler.java]: `false` is the default value of a `boolean` variable, no need to initialize explicitly. But I don't mind that much.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddLogCategory.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+import org.wildfly.extras.creaper.core.online.OnlineCommandContext;
+import org.wildfly.extras.creaper.core.online.operations.Address;
+import org.wildfly.extras.creaper.core.online.operations.Batch;
+import org.wildfly.extras.creaper.core.online.operations.OperationException;
+import org.wildfly.extras.creaper.core.online.operations.Operations;
+import org.wildfly.extras.creaper.core.online.operations.Values;
+
+/**
+ * <p>
+ * Command adds log category.
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddLogCategory.java]: This comment only replicates the class name. If there's nothing more to say than the class name, just avoid the comment completely.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddLogCategory.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+ * </p>
+ */
+
+public class AddLogCategory extends ManipulateLogCategory {
+
+    private boolean replaceExisting;
+
+    protected AddLogCategory(Builder builder) {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddLogCategory.java]: `protected`?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddLogCategory.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+                .andOptional("level", level == null ? null : level.value())
+                .andOptional("filter-spec", filter)
+                .andOptional("use-parent-handlers", useParentHandler)
+                .andListOptional(String.class, "handlers", handlers);
+
+
+        Batch batch = new Batch();
+        batch.add(loggerAddress, values);
+
+        ops.batch(batch);
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddLogCategory.java]: Again, why a batch with a single operation? If you want, you can use the batch to do `replaceExisting` in a transactional fashion, but you're not doing it right now, so the batch is useless.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddPeriodicHandler.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+                throw new IllegalArgumentException("file can not be null!");
+            }
+            super.validate();
+        }
+
+        /**
+         * Set true if replacing of existing node is needed.
+         * Default value is false.
+         */
+        public Builder setReplaceExisting(final boolean value) {
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/AddPeriodicHandler.java]: Avoid `final` on parameters unless needed.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/HandlerType.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package org.wildfly.extras.creaper.commands.logging;
+
+/**
+ * @author Ivan Straka istraka@redhat.com
+ */
+public enum HandlerType {
+
+    CONSOLE("console-handler"),
+    PERIODIC("periodic-rotating-file-handler");
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/HandlerType.java]: There are at least two important properties of this log handler -- that it's file-based and that it uses periodic log rotation. I'd call it `FILE_ROTATING` or something like that.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/HandlerType.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    private final String val;
+
+    /**
+     * @param val
+     */
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/HandlerType.java]: ?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+    DEBUG("DEBUG"),
+    FINE("FINE"),
+    CONFIG("CONFIG"),
+    INFO("INFO"),
+    WARN("WARN"),
+    WARNING("WARNING"),
+    ERROR("ERROR"),
+    SEVERE("SEVERE"),
+    FATAL("FATAL"),
+    OFF("OFF");
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java]: Make sure these are sorted. I think `FINER` sorts _after_ `TRACE` (actually, I think that `FINEST` and `TRACE` are synonyms, `FINER` and `DEBUG` too, but please confirm first).

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+    private final String val;
+
+    /**
+     * @param val
+     */
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java]: ?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        this.val = val;
+    }
+
+    public String value() {
+        return val;
+    }
+};
+
+
+
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java]: Why the empty lines?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/ManipulateConsoleHandler.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        public void validate() {
+            if (patternFormatter != null && namedFormatter != null) {
+                throw new IllegalStateException("You can not define both pattern formater and named formatter.");
+            }
+        }
+
+        /**
+         * @param val
+         * @return
+         */
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/ManipulateConsoleHandler.java]: ?

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/ManipulateConsoleHandler.java [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+            if (filter == null) {
+                throw new IllegalArgumentException("filter can not be null.");
+            }
+            this.filter = filter;
+            return (THIS) this;
+        }
+
+        /**
+         * You can use PATTERN. This should be configured already.
+         */
```
Comment (Ladicek) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/ManipulateConsoleHandler.java]: What does this comment mean?

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandlerOfflineTest.java [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+    @Test(expected = CommandFailedException.class)
+    public void replaceExisting2() throws Exception {
+        File cfg = tmp.newFile("xmlTransform.xml");
+        String loggingXmlOriginal = String.format(HANDLER_ORIGINAL,
+                "            <console-handler name=\"consolehandler\" autoflush=\"false\" enabled=\"false\">\n"
+                        + "                <level name=\"FINEST\"/>\n"
+                        + "                <filter-spec value=\"match(&quot;filter*&quot;)\"/>"
+                        + "                <encoding value=\"UTF-8\"/>\n"
+                        + "                <target name=\"System.out\"/>\n"
+                        + "            </console-handler>\n"
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/logging/AddConsoleHandlerOfflineTest.java]: Formatting.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/logging/ChangeConsoleHandlerOfflineTest.java [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+        client.apply(changeConsoleHandler);
+        assertXmlIdentical(HANDLER_CHANGE_EXPECTED, Files.toString(cfg, Charsets.UTF_8));
+    }
+
+    @Test(expected = CommandFailedException.class)
+    public void nonExisting() throws Exception {
+        File cfg = tmp.newFile("xmlTransform.xml");
+        String loggingXmlOriginal = String.format(HANDLER_ORIGINAL,
+                "            <console-handler name=\"consolehandler\">\n"
+                        + "            </console-handler>\n"
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/logging/ChangeConsoleHandlerOfflineTest.java]: Formatting.

File: testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/logging/RemoveLogCategoryOnlineTest.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    public void addLogger() throws Exception {
+        AddLogCategory addLogCategory = new AddLogCategory.Builder(TEST_LOGGER_NAME)
+                .level(Level.OFF)
+                .filter("match(\"filter\")")
+                .setUseParentHandler(true)
+                .build();
+
+        client.apply(addLogCategory);
+
+        assertTrue("logger wasn't created", ops.exists(TEST_LOGGER_ADDRESS));
```
Comment (Ladicek) [inline testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/logging/RemoveLogCategoryOnlineTest.java]: Please use "should"-style assertion messages. Here e.g. `"Logger should be created"`.

**Comment by Ladicek** [comment]
I like the comprehensive test coverage, good job with that!

Couple objections:
- I don't think that you ran Checkstyle on this. Please use `mvn clean verify`.
- The `Manipulate...` classes: I'd prefer a name starting with `Abstract...` and avoiding the `Manipulate` thing. Also, do they have to be public?
- The subclasses of `Manipulate...` -- why not `final`?

Also, the PR is huge, which is terrible for review. In the future, a well-structured series of PRs will lead to easier review and faster merging. This one will take a lot of time to process.

**Comment by Ladicek** [comment]
Also, `has been implemented` in the commit message -- please avoid that. It's clear from the very existence of the commit that something has been implemented.

I'd prefer something like `add basic commands for configuring logging subsystem`.

File: commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java (hunk 10 above)
Comment (istraka) [inline commands/src/main/java/org/wildfly/extras/creaper/commands/logging/Level.java]: `DEBUG` is same as `FINER` and `FINE` according to http://docs.jboss.org/mobicents/sip-presence/1.0.0.FINAL/html/logging_traces_and_alarms.html#tracers_vs_log4j

**Comment by istraka** [comment]
Issues have been fixed and first commit was renamed. Checkstyle was ran successffuly.

**Comment by Ladicek** [comment]
Besides the comments in the commit, I think we need to fix naming a bit.

First, `ManipulateConsoleHandler` wasn't very descriptive, but `AbstractConsoleHandler` isn't much better (it isn't really an abstract implementation of a console handler, but a base class for _commands_ that relate to console handlers). Since these classes are not `public`, I think we can call them `AbstractConsoleHandlerCommand`.

Second, you only renamed one of the commands dealing with the periodic rotating file handler; the rest are still called the old way. Also, I think that `AddPeriodicRotatingFileHandler` is too long and it isn't immediately obvious that it is related to logging. I think restructuring a bit would be best here. What about a bit of nesting? E.g. `static` methods `AddLogHandler.periodicRotatingFile()` or just `AddLogHandler.rotatingFile()`?

Note that I'd like to _discuss_ this first. Let's get to an agreement and only then can we implement changes.

**Comment by istraka** [comment]
Design have been changed a little bit.

**Comment by Ladicek** [comment]
I did a bit of a refactoring :-), see https://github.com/Ladicek/creaper/commits/logging

In the end, I moved the code to the common scheme (publicly available `Add|Change|Remove*` commands) with a single class `Logging` serving as a more convenient entrypoint. (This looks like a useful pattern that could be adopted e.g. for datasources.)

I'm not entirely sure that the combination of private/public classes makes the commands usable outside of Creaper, but the tests pass just fine. (I had to do a small fix though: #38.)

**Comment by Ladicek** [comment]
I have rebased and reviewed again the code that I'd like to merge: https://github.com/Ladicek/creaper/commit/4a2e3d3269

Could you please take a look as well?

(Also, I changed your e-mail in the commit. Please set `~/.gitconfig` correctly. Preferrably, add your Red Hat e-mail to your GitHub account, too.)

**Comment by istraka** [comment]
I find it much better for users. LGTM.

**Comment by Ladicek** [comment]
Superseded by #47.

**Pull request closed without merging**